ex5/y_ogawa/main.py

```diff
@@ -0,0 +1,348 @@
+"""GMMを用いたデータフィッティングを行う."""
+
+import argparse
+
+import matplotlib.pyplot as plt
+import matplotlib.ticker as ticker
+import numpy as np
+from scipy.stats import multivariate_normal
+
+import ex4
+
+
+def parse_args():
+    """引数の取得を行う.
+
+    filename : 読み込むファイル名
+    n_cluster : クラスター数
+    """
+    parser = argparse.ArgumentParser(description="GMMを用いたマッピングを行う")
+    parser.add_argument("--filename", type=str, required=True, help="name of file")
+    parser.add_argument(
+        "--n_components", type=int, required=True, help="number of component"
+    )
+    return parser.parse_args()
+
+
+def random_initialize_para(data: np.ndarray, n_components: int) -> np.ndarray:
+    """ランダムでパラメータを初期化する.
+
+    Args:
+        data (np.ndarray): データ
+        n_components (int): 混合数
+
+    Returns:
+        weights: 混合係数
+        means: 平均
+        covariances: 共分散行列
+    """
+    n_features = data.shape[1]
+
+    # 混合係数をランダムに初期化 (正規化する)
+    weights = np.random.rand(n_components)
+    weights /= weights.sum()
+
+    # 平均値をランダムに初期化
+    means = np.random.rand(n_components, n_features) * (
+        data.max(axis=0) - data.min(axis=0)
+    ) + data.min(axis=0)
```

Comment:
```suggestion
    kmeans = KMeans(n_clusters=n_components, random_state=0).fit(data)
    means = kmeans.cluster_centers_
```
ランダムに初期化していたことが原因でエラーが発生していると考察していましたが、その通りだと思われます。
具体的には、平均値の初期化位置が偏っていたりすると、ほとんど要素を持たないクラスが生じるなどデータ不均衡が発生し、共分散行列の固有値が限りなくゼロに近づいていきます。

対策として、[k-means](https://qiita.com/g-k/items/0d5d22a12a4507ecbf11)法によってある程度データをクラスタリングし、各クラスの中心値（重心）をそのまま初期値とする方法が主流のようです。

k-means法を用いて初期化することにより初期値の偏りが生じにくくなるので、エラーが発生しにくくなります。
逆にk-means法を用いたとしても、データが明らかに偏ったりしているとエラーが発生することもあるようです。